src/mds/Server.cc

```diff
@@ -436,11 +436,13 @@ void Server::reclaim_session(Session *session, const cref_t<MClientReclaim> &m)
     return;
   }
 
-  unsigned flags = m->get_flags();
-  if (flags != CEPH_RECLAIM_RESET) { // currently only support reset
-    dout(10) << __func__ << " unsupported flags" << dendl;
+  unsigned flag = m->get_flags();
+  if (flag != CEPH_RECLAIM_RESET) { // currently only support reset
+    dout(10) << __func__ << " unsupported flag" << dendl;
     reply->set_result(-CEPHFS_EOPNOTSUPP);
     mds->send_message_client(reply, session);
+    CachedStackStringStream css;
+    mds->evict_client(session->get_client().v, false, true, *css, nullptr);
```

Comment:
there is no need to evict the client if the MDS has always sent `EOPNOTSUPP`.

Reply:
I was thinking that maybe we should evict the client here instead of sending `EOPNOTSUPP` then?

Reply:
No. It's better for the client to handle compatibility with a genuine error rather than requiring a feature bit for every new flag we create.

Reply:
@batrick nope, no way. We need to be able to query for appropriate behaviors without just trying an op and observing if it succeeds, or we can get into untenable states due to the way our distributed cache is maintained across both MClientOps and MClientCaps changes.

Reply:
I don't see how a client learning a flag for reclaim is invalid via EOPNOTSUPP and then trying a more vanilla reclaim is untenable. This is how it works for e.g. with the kernel and I don't agree a distributed cache/system changes this.

Evicting the client should be a last resort.

Reply:
@batrick So when the MDS sends the client `EOPNOTSUPP`, it would print that message and move on? Or what else would it do?

Reply:
Okay, so won't it be better to be strict with client and evict it on it's first try of sending unknown reclaim mode?

Reply:
No. Let the client retry with a different mode.

Reply:
Also, the result should have been `-CEPHFS_EINVAL` rather than `-CEPHFS_EOPNOTSUPP`.

Reply:
Okay, I'll make that change.

Reply:
BTW isn't this an OP? Or is it an argument?

Reply:
It's a flag. Generally, `-EINVAL` is preferred for unknown/invalid flags to a command. See for example `open(2)`.

Reply:
Got it, thanks!

Reply:
> I don't see how a client learning a flag for reclaim is invalid via EOPNOTSUPP and then trying a more vanilla reclaim is untenable. This is how it works for e.g. with the kernel and I don't agree a distributed cache/system changes this.

We split state updates amongst MDS client requests like this, and MDS caps updates. So in the general case,
1) the end user does a syscall
2) the kclient makes changes to its local cache to reflect that change
3) the kclient sends out an MClientRequest as needed
4) *at some other time, perhaps before getting a reply to (3)* the kclient sends out an MClientCaps which reflects the user change.

If we send an error back to the kclient and it can't handle it, what happens with the state then? If it's a new end-user op, it can just return EOPNOTSUPP through the syscall, but we've already reflected some of the changes (like, possibly, size) to the rest of the cluster through the MClientCaps.

Now, we can in some specific cases know that this isn't an issue. But we can't know that for a general case, and we can't predict what might happen in the future even with something like client reclaim (what if we, in the future, implement a version that allows an existing client to take over the state from one that dies, without losing its own state? An EOPNOTSUPP would *not* suffice there!).

> Evicting the client should be a last resort.
Agreed. But the client sending ops the cluster can't handle triggers that condition.

Reply:
You could certainly design a distributed FS where just returning EOPNOTSUP is the way to tell the client about existing features. But CephFS wasn't designed that way — we'd need to implement inode projection on the client as well as the server side, or make sure we always hold updates to the "real" inode until after requests are sent off. And more generally any kind of deferred or cached update would become impossible.

Reply:
> > I don't see how a client learning a flag for reclaim is invalid via EOPNOTSUPP and then trying a more vanilla reclaim is untenable. This is how it works for e.g. with the kernel and I don't agree a distributed cache/system changes this.
> 
> We split state updates amongst MDS client requests like this, and MDS caps updates. So in the general case,
> 
>     1. the end user does a syscall
> 
>     2. the kclient makes changes to its local cache to reflect that change
> 
>     3. the kclient sends out an MClientRequest as needed
> 
>     4. _at some other time, perhaps before getting a reply to (3)_ the kclient sends out an MClientCaps which reflects the user change.

I'm not aware of an instance of this (4). All requests by both the kernel and fuse client synchronously wait for the result (in the thread which makes the call). `MClientCaps` state updates would reflect changes for which the client already has rights to mutate metadata.

> If we send an error back to the kclient and it can't handle it, what happens with the state then? If it's a new end-user op, it can just return EOPNOTSUPP through the syscall, but we've already reflected some of the changes (like, possibly, size) to the rest of the cluster through the MClientCaps.

You can't modify client cache state depending on the result of an in-flight MDS request (`MClientRequest`). Please give an example.

>  (what if we, in the future, implement a version that allows an existing client to take over the state from one that dies, without losing its own state? An EOPNOTSUPP would _not_ suffice there!).

Why not? The client sees it cannot steal state from the other session and then issues a new `MClientReclaim` with flag `CEPH_RECLAIM_RESET` instead.

Reply:
> > I don't see how a client learning a flag for reclaim is invalid via EOPNOTSUPP and then trying a more vanilla reclaim is untenable. This is how it works for e.g. with the kernel and I don't agree a distributed cache/system changes this.
> 
> We split state updates amongst MDS client requests like this, and MDS caps updates. So in the general case,
> 
> 1. the end user does a syscall
> 2. the kclient makes changes to its local cache to reflect that change
> 3. the kclient sends out an MClientRequest as needed
> 4. _at some other time, perhaps before getting a reply to (3)_ the kclient sends out an MClientCaps which reflects the user change.
> 
> If we send an error back to the kclient and it can't handle it, what happens with the state then? If it's a new end-user op, it can just return EOPNOTSUPP through the syscall, but we've already reflected some of the changes (like, possibly, size) to the rest of the cluster through the MClientCaps.

Does MClientCaps work in tandem with MClientRequest? If yes then most probably changes reflected by it could be backtracked right? Or no?

> Now, we can in some specific cases know that this isn't an issue. But we can't know that for a general case, and we can't predict what might happen in the future even with something like client reclaim (what if we, in the future, implement a version that allows an existing client to take over the state from one that dies, without losing its own state? An EOPNOTSUPP would _not_ suffice there!).
> 
> > Evicting the client should be a last resort.
> 
> Agreed. But the client sending ops the cluster can't handle triggers that condition.

Reply:
> > > I don't see how a client learning a flag for reclaim is invalid via EOPNOTSUPP and then trying a more vanilla reclaim is untenable. This is how it works for e.g. with the kernel and I don't agree a distributed cache/system changes this.
> > 
> > 
> > We split state updates amongst MDS client requests like this, and MDS caps updates. So in the general case,
> > ```
> > 1. the end user does a syscall
> > 
> > 2. the kclient makes changes to its local cache to reflect that change
> > 
> > 3. the kclient sends out an MClientRequest as needed
> > 
> > 4. _at some other time, perhaps before getting a reply to (3)_ the kclient sends out an MClientCaps which reflects the user change.
> > ```
> 
> I'm not aware of an instance of this (4). All requests by both the kernel and fuse client synchronously wait for the result (in the thread which makes the call). `MClientCaps` state updates would reflect changes for which the client already has rights to mutate metadata.
> 
> > If we send an error back to the kclient and it can't handle it, what happens with the state then? If it's a new end-user op, it can just return EOPNOTSUPP through the syscall, but we've already reflected some of the changes (like, possibly, size) to the rest of the cluster through the MClientCaps.
> 
> You can't modify client cache state depending on the result of an in-flight MDS request (`MClientRequest`). Please give an example.
> 
> > (what if we, in the future, implement a version that allows an existing client to take over the state from one that dies, without losing its own state? An EOPNOTSUPP would _not_ suffice there!).
> 
> Why not? The client sees it cannot steal state from the other session and then issues a new `MClientReclaim` with flag `CEPH_RECLAIM_RESET` instead.

If it's the same buggy client trying to issue a new Reclaim with the same troubling flag, won't that be a bit waste of resource? Also what happens when the Client is aimed to make troubles by continuously sending invalid flags? Will it keep on giving the Client another chance to send the correct flag?

Reply:
> > I don't see how a client learning a flag for reclaim is invalid via EOPNOTSUPP and then trying a more vanilla reclaim is untenable. This is how it works for e.g. with the kernel and I don't agree a distributed cache/system changes this.
> 
> We split state updates amongst MDS client requests like this, and MDS caps updates. So in the general case,
> 
> 1. the end user does a syscall
> 2. the kclient makes changes to its local cache to reflect that change
> 3. the kclient sends out an MClientRequest as needed
> 4. _at some other time, perhaps before getting a reply to (3)_ the kclient sends out an MClientCaps which reflects the user change.
> 

There already has a bug like this, please see https://github.com/ceph/ceph/pull/46357. If the xattrs key/value exceeding the limits and at the same time the `Xx` has been issued, the clients will cache that change locally and return success. But when flushing the caps back to MDS it will fail.


> If we send an error back to the kclient and it can't handle it, what happens with the state then? If it's a new end-user op, it can just return EOPNOTSUPP through the syscall, but we've already reflected some of the changes (like, possibly, size) to the rest of the cluster through the MClientCaps.
>

We need to properly handle this case. As discussed before in that PR or kernel patch for the xattr limits issue the kclient can be sure that the MDS won't fail and then it can cache the new change. Or just do a sync request immediately instead to let the MDS to decide how to do it properly.

Reply:
> There already has a bug like this, please see #46357. If the xattrs key/value exceeding the limits and at the same time the Xx has been issued, the clients will cache that change locally and return success. But when flushing the caps back to MDS it will fail.

But this is not the same situation. The kernel is buffering xattr mutations but no synchronous I/O is necessary. It's just a cap flush.

Reply:
> I don't see how a client learning a flag for reclaim is invalid via EOPNOTSUPP and then trying a more vanilla reclaim is untenable. This is how it works for e.g. with the kernel and I don't agree a distributed cache/system changes this.

I agree, at least in the above case (for state reclaim), its probably OK for the client to switch using vanilla reclaim mechanism.

@dparmar18 has an audit report on various `ceph_abort()` calls in the MDS. We should carefully see which of those calls _really_ need to be switched to evicting the client vs returning errno back to the client.

WDYT?

Reply:
maybe we should wait for such scenarios to occur and then go through all those abort calls